### PR TITLE
pwru: init at 1.0.0

### PR DESCRIPTION
**Pull request by gaelreyrol**
###### Description of changes

This PR adds pwru package: https://github.com/cilium/pwru

Closes #239946
Closes #224886

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by gaelreyrol** [comment]
Oops It is some kind a duplicate of @SuperSandro2000's PR https://github.com/NixOS/nixpkgs/pull/224886/ but with the latest version.
This derivation handles the `pcap-filter` feature.

File: pkgs/tools/networking/pwru/default.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    hash = "sha256-9UMuHDPDL8nPdL6U+BHHSXvd0iDUigVOa5Tel7n5L4Q=";
+  };
+
+  patches = [
+    ./go_generate_cflags.patch
+  ];
+
+  vendorHash = null;
+
+  CGO_ENABLED = 1;
```
Comment (AndersonTorres) [inline pkgs/tools/networking/pwru/default.nix:31]:
```suggestion
  env.CGO_ENABLED = "1";
```

Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/default.nix:31]: For my personal knowledge, why should it be in env? I don't find any derivation using this declaration.

Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/default.nix:31]: When I try your suggestion I also get this error: `error: The ‘env’ attribute set cannot contain any attributes passed to derivation. The following attributes are overlapping: CGO_ENABLED`. The buildGoModule has an explicit attribute for `CGO_ENABLED`: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/go/module.nix#L35

Comment (AndersonTorres) [reply pkgs/tools/networking/pwru/default.nix:31]: Sorry, I didn't noticed. I thought it was an environment variable, not a parameter.

**Review by AndersonTorres — APPROVED** [review]

**Comment by gaelreyrol** [comment]
I am trying to make it work on aarch64-linux.

File: pkgs/tools/networking/pwru/default.nix [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    env TARGET_GOARCH=${goArch} \
+      go generate
+
+    mv vendor/github.com/cilium/ebpf/internal/cpu.go{.bak,}
+  '';
+
+  ldflags = [
+    "-w"
+    "-s"
+    "-linkmode=external"
```
Comment (Mindavi) [inline pkgs/tools/networking/pwru/default.nix:92]: Why is this needed? Upstream doesn't appear to use it

Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/default.nix:92]: If I don't use it I get these errors on amd64 & arm64:
```
runtime.gcdata: missing Go type information for global symbol .dynsym: size 72
runtime/cgo(.text): relocation target stderr not defined
runtime/cgo(.text): relocation target fwrite not defined
runtime/cgo(.text): relocation target __vfprintf_chk not defined
runtime/cgo(.text): relocation target fputc not defined
runtime/cgo(.text): relocation target abort not defined
runtime/cgo(.text): relocation target pthread_mutex_lock not defined
runtime/cgo(.text): relocation target pthread_cond_wait not defined
runtime/cgo(.text): relocation target pthread_mutex_unlock not defined
runtime/cgo(.text): relocation target pthread_cond_broadcast not defined
runtime/cgo(.text): relocation target pthread_create not defined
runtime/cgo(.text): relocation target nanosleep not defined
runtime/cgo(.text): relocation target pthread_detach not defined
runtime/cgo(.text): relocation target strerror not defined
runtime/cgo(.text): relocation target stderr not defined
runtime/cgo(.text): relocation target __fprintf_chk not defined
runtime/cgo(.text): relocation target abort not defined
runtime/cgo(.text): relocation target free not defined
runtime/cgo(.text): relocation target malloc not defined
runtime/cgo(.text): relocation target pthread_attr_init not defined
runtime/cgo(.text): relocation target pthread_attr_getstacksize not defined
```

File: pkgs/tools/networking/pwru/fix-pcap-cmake.patch [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+-        install_manpage_symlink(pcap_list_tstamp_types.3pcap pcap_free_tstamp_types.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_loop.3pcap pcap_dispatch.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_major_version.3pcap pcap_minor_version.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_next_ex.3pcap pcap_next.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_open_dead.3pcap pcap_open_dead_with_tstamp_precision.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_open_offline.3pcap pcap_open_offline_with_tstamp_precision.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_open_offline.3pcap pcap_fopen_offline.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_open_offline.3pcap pcap_fopen_offline_with_tstamp_precision.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_tstamp_type_val_to_name.3pcap pcap_tstamp_type_val_to_description.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
+-        install_manpage_symlink(pcap_setnonblock.3pcap pcap_getnonblock.3pcap ${CMAKE_INSTALL_MANDIR}/man3)
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/pwru/fix-pcap-cmake.patch:34]: Why don't we want the man pages?

File: pkgs/tools/networking/pwru/default.nix [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  depsBuildBuild = [ buildPackages.stdenv.cc ];
+
+  buildInputs = [ stdenv.cc.libc.static libpcap ];
+
+  preBuild = ''
+    echo "0-3" > system-cpu-possible
+    sed -i.bak vendor/github.com/cilium/ebpf/internal/cpu.go \
+      -e "s|/sys/devices/system/cpu/possible|$(realpath system-cpu-possible)|"
+
+    env TARGET_GOARCH=${goArch} \
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/pwru/default.nix:83]: Do we really need to set go arch? I got away with not setting it

File: pkgs/tools/networking/pwru/default.nix [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  pname = "pwru";
+  version = "1.0.0";
+  src = fetchFromGitHub {
+    owner = "cilium";
+    repo = "pwru";
+    rev = "v${version}";
+    hash = "sha256-9UMuHDPDL8nPdL6U+BHHSXvd0iDUigVOa5Tel7n5L4Q=";
+  };
+
+  libpcap = stdenv.mkDerivation {
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/pwru/default.nix:31]: Why do we use their vendored libpcap? We can't just use ours?

File: pkgs/tools/networking/pwru/default.nix (hunk 4 above)
Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/default.nix:83]: I get this error if I don't set it:
```
       > Generating for
       > Supported targets:
       >    bpf
       >    bpfel
       >  bpfeb
       >  386
       >    amd64
       >  arm
       >    arm64
       >  arm64be
       >        armbe
       >  ppc64
       >  ppc64le
       >        s390
       >   s390x
       >  sparc
       >  sparc64
       >
       > Error: "": unsupported target
```

File: pkgs/tools/networking/pwru/default.nix (hunk 5 above)
Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/default.nix:31]: Because those flags are applied:
```
    cmakeFlags = [
      "-DBUILD_SHARED_LIBS=OFF"
      "-DDISABLE_LINUX_USBMON=ON"
      "-DDISABLE_BLUETOOTH=ON"
      "-DDISABLE_NETMAP=ON"
      "-DDISABLE_DBUS=ON"
      "-DBUILD_WITH_LIBNL=OFF"
    ];
```

File: pkgs/tools/networking/pwru/fix-pcap-cmake.patch (hunk 3 above)
Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/fix-pcap-cmake.patch:34]: Because it seems the man path is wrongly build:
```
CMake Error at cmake_install.cmake:176 (message):
  Could not create symbolic link from
  /nix/store/jn69b316l09rmci02jh4h7rmps2f2mqq-pwru-pcap-1.0.0//nix/store/jn69b316l09rmci02jh4h7rmps2f2mqq-pwru-pcap-1.0.0/share/man/man3/pcap_d
atalink_val_to_name.3pcap
  to pcap_datalink_val_to_description.3pcap
```

Comment (SuperSandro2000) [reply pkgs/tools/networking/pwru/fix-pcap-cmake.patch:34]: Can you try dropping CMAKE_INSTALL_MANDIR from the install path?

File: pkgs/tools/networking/pwru/default.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/tools/networking/pwru/default.nix:31]: It shouldn't be a problem if we supply a lib with more features, right?

Comment (gaelreyrol) [reply pkgs/tools/networking/pwru/default.nix:31]: Because of the static build it seems that if libpcap is not compiled that way, when the build of  Go modules occurs in static mode it fails to find libnl.

**Comment by AndersonTorres** [comment]
ping

**Comment by mfrw** [comment]
Result of `nixpkgs-review pr 243638` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pwru</li>
  </ul>
</details>
